File: backend/app/booking/service.py
from sqlalchemy.ext.asyncio import AsyncSession
from app.booking.repository import BookingRepository
from app.booking.schemas import BookingCreate, BookingRead
from app.websocket.manager import room_manager


class BookingService:
    def __init__(self, db: AsyncSession):
        self.db = db
        self.repository = BookingRepository(db)

    async def book_unit(self, data: BookingCreate) -> BookingRead:
        booking = await self.repository.create_booking_atomic(
            unit_id=data.unitId,
            customer_name=data.customerName,
            phone=data.phone
        )
        
        # Broadcast real-time booking event to all paired session clients
        await room_manager.broadcast_unit_booked(
            session_id=data.sessionId,
            unit_id=data.unitId,
            customer_name=data.customerName
        )

        return BookingRead.model_validate(booking)

File: kiosk-backend/src/modules/booking/service.ts
import { BookingRepository, BookingRow } from './repository';
import { roomManager } from '../websocket/manager';

export interface BookingPayload {
  unitId: number;
  customerName: string;
  phone: string;
  sessionId?: string;
}

export class BookingService {
  private repository: BookingRepository;

  constructor() {
    this.repository = new BookingRepository();
  }

  async bookUnit(data: BookingPayload): Promise<BookingRow> {
    const booking = await this.repository.createBookingAtomic(
      data.unitId,
      data.customerName,
      data.phone
    );

    const sessionId = data.sessionId || 'default';
    await roomManager.broadcastUnitBooked(sessionId, data.unitId, data.customerName);

    return booking;
  }
}
